### ISSUE TITLE
Ivy ignores components needed for dynamic component loader

### ISSUE THREAD

**Issue by TheSlimvReal** [angular/angular#45387]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

In my application I am using a decorator `@DynamicComponent` to marks components that can later be rendered using `viewContainerRef.createComponent(...)` (see the guide for [dynamic component loading](https://angular.io/guide/dynamic-component-loader)).

```javascript
@DynamicComponent('hello')
@Component({ template: `<h1>Hello Angular</h1>` })
export class HelloComponent {}
```
The decorator registers the component with a string on a static map.
```javascript
export function DynamicComponent(name: string) {
  return (ctor) => {
    componentRegistry.set(name, ctor);
  };
}
export const componentRegistry = new Map<string, any>();
```
Later the same string can be used to find the component and to load/create it dynamically.
```javascript
@Component({ template: `<ng-template anchor></ng-template>` })
export class AppComponent implements OnInit {
  @ViewChild(AnchorDirective, { static: true }) anchorHost!: AnchorDirective;
  ngOnInit() {
    const viewContainerRef = this.anchorHost.viewContainerRef;
    viewContainerRef.clear();
    const component = componentRegistry.get('hello');
    viewContainerRef.createComponent(component);
  }
}
```
This works fine in development mode.

However, in production (`ng serve --configuration production`) the decorators are never executed because Ivy believes the components are not used/needed.
This is because the dynamic components are only used in the modules where they are declared.
The moment I write some dummy code using the dynamic component (e.g. `new HelloComponent()`) everything works fine again.

I provided a stackblitz below where this is implemented.
In stackblitz everything is working fine because it is started in development mode.
To run it in production mode download the stackblitz example and run `npm install --force && npm start`.

### Please provide a link to a minimal reproduction of the bug

https://angular-ivy-ypsare.stackblitz.io

### Please provide the exception or error you saw

```true
ERROR TypeError: Cannot read properties of undefined (reading 'ɵcmp')
```
```


### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 13.3.0
Node: 16.14.2
Package Manager: npm 8.5.0
OS: win32 x64

Angular: 13.3.0
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1303.0
@angular-devkit/build-angular   13.3.0
@angular-devkit/core            13.3.0
@angular-devkit/schematics      13.3.0
@schematics/angular             13.3.0
rxjs                            7.5.5
typescript
```

### Anything else?

I believe this was the initial reason why the `entryComponents` module option existed which are now deprecated and considered unnecessary.
I believe something like this is necessary again in order to mark components that should not be ignored by Ivy just because it cannot find a direct reference to it in the source code.
Without any changes I believe the loading & creating dynamic component is not working as expected as it requires the components to be at least once loaded/created **non**-dynamically.

I would be thankful for any hint on how to get this thing running in production even if it would mean turning off this unused-component cleanup done by Ivy.

**Comment by JoostK** [comment]
This is working as designed. In optimized builds any symbols that are not referenced are tree-shaken away. Some more details are available in #35314.

There is no need for an alternative for `entryComponents`, as you can easily create hard references using a static field on a class (instead of listing the components in the `NgModule.entryComponents` metadata).

**Comment by TheSlimvReal** [comment]
Thank you for the details and the reference to the older issue. Good explanation and I will go with the static references even though it feels like a very non-Angular approach.
